### PR TITLE
Add additional services specific controls. Closes #29

### LINKED ISSUE #29: Add additional services specific controls
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Priyanka-Chatterjee-2000**
### Checklist
**Newly Added Controls and queries**

1. [x] artifact_registry_repository_not_publicly_accessible
2. [x] artifact_registry_repository_encrypted_with_kms_cmk
3. [x] bigquery_table_not_publicly_accessible
4. [x] bigquery_instance_encrypted_with_kms_cmk
5. [x] compute_security_policy_prevent_message_lookup
6. [x] cloudbuild_workers_use_private_ip
7. [x] cloudfunction_not_publicly_accessible
8. [x] pubsub_topic_encrypted_with_kms_cmk
9. [x] sql_instance_using_latest_major_database_version
10. [x] storage_bucket_logging_enabled
11. [x] storage_bucket_self_logging_disabled
12. [x] storage_bucket_versioning_enabled
13. [x] datafusioninstance_not_publicly_accessible
14. [x] datafusioninstance_stackdriver_logging_enabled
15. [x] datafusioninstance_stackdriver_monitoring_enabled
16. [x] dataflow_encrypted_with_kms_cmk
17. [x] dataflow_job_not_publicly_accessible
18. [x] dataproc_cluster_encrypted_with_kms_cmk
19. [x] dataproc_cluster_not_publicly_accessible
20. [x] dataproc_cluster_public_ip_disabled
21. [x] cloudrun_not_publicly_accessible
22. [x] kubernetes_cluster_alias_ip_range_enabled
23. [x] kubernetes_cluster_binary_auth_enabled
24. [x] kubernetes_cluster_client_certificate_authentication_disabled
25. [x] kubernetes_cluster_stackdriver_logging_enabled
26. [x] kubernetes_cluster_shielded_nodes_enabled
27. [x] kubernetes_cluster_intranode_visibility_enabled
28. [x] kubernetes_cluster_shielded_node_integrity_monitoring_enabled
29. [x] kubernetes_cluster_resource_label_configured
30. [x] kubernetes_cluster_authenticator_group_configured
31. [x] kubernetes_cluster_master_authorized_network_enabled
32. [x] kubernetes_cluster_metadata_server_enabled
33. [x] kubernetes_cluster_stackdriver_monitoring_enabled
34. [x] kubernetes_cluster_control_plane_restrict_public_access
35. [x] kubernetes_cluster_release_channel_configured
36. [x] kubernetes_cluster_shielded_node_secure_boot_enabled
37. [x] kubernetes_cluster_cos_node_image
38. [x] sql_instance_postgresql_pgaudit_enabled
39. [x] sql_instance_postgresql_log_min_error_statement_flag_set
40. [x] sql_instance_postgresql_log_min_messages_flag_set
41. [x] sql_instance_postgresql_log_statement_flag_set
42. [x] sql_instance_publicly_accessible
43. [x] sql_instance_sql_with_no_public_ip
44. [x] compute_instance_boot_disk_encryption_enabled
45. [x] compute_firewall_allow_ftp_port_20_ingress
46. [x] compute_firewall_allow_ftp_port_21_ingress
47. [x] compute_firewall_allow_ssh_port_22_ingress
48. [x] compute_firewall_allow_mysql_port_3306_ingress
49. [x] compute_firewall_allow_rdp_port_3389_ingress
50. [x] compute_firewall_allow_http_port_80_ingress
51. [x] compute_compute_project_metadata_oslogin_enabled
52. [x] compute_load_balancer_use_secure_ssl_cipher
53. [x] iam_folder_use_basic_role
54. [x] iam_folder_impersonation_role
55. [x] iam_folder_use_default_service_role
56. [x] kms_key_not_publicly_accessible
57. [x] kms_key_prevent_destroy_enabled
58. [x] iam_organisation_use_basic_role
59. [x] iam_organisation_impersonation_role
60. [x] iam_organization_use_default_service_role
61. [x] iam_service_account_no_admin_priviledge
62. [x] iam_project_use_basic_role
63. [x] project_no_default_network
64. [x] iam_project_impersonation_role
65. [x] iam_project_use_default_service_role
66. [x] iam_project_no_service_account_token_creator_role
67. [x] storage_bucket_public_access_prevention_enforced
68. [x] compute_subnetwork_private_ipv6_google_access
69. [x] redis_instance_auth_enabled
70. [x] redis_instance_encryption_in_transit_enabled
71. [x] pubsub_topic_repository_not_publicly_accessible
72. [x] spanner_database_encrypted_with_kms_cmk
73. [x] vertex_ai_encrypted_with_cmk
74. [x] vertex_ai_notebook_instance_restrict_public_access

**Query Updated**

1. [x] bigquery_dataset_not_publicly_accessible
2. [x] compute_instance_block_project_wide_ssh_enabled
3. [x] compute_instance_with_no_default_service_account
4. [x] compute_instance_ip_forwarding_disabled
5. [x] compute_subnetwork_flow_log_enabled

### CODE REVIEW

File: query/sql.sp [hunk 1]
```diff
@@ -793,3 +793,168 @@ query "sql_instance_automated_backups_enabled" {
...
+query "sql_instance_using_latest_major_database_version" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments ->> 'database_version') in ('POSTGRES_15','MYSQL_8_0','SQLSERVER_2019_STANDARD','SQLSERVER_2019_WEB','SQLSERVER_2019_ENTERPRISE','SQLSERVER_2019_EXPRESS') then 'ok'
+        else 'alarm'
+      end as status,
+      name || case
+        when (arguments ->> 'database_version') in ('POSTGRES_15','MYSQL_8_0','SQLSERVER_2019_STANDARD','SQLSERVER_2019_WEB','SQLSERVER_2019_ENTERPRISE','SQLSERVER_2019_EXPRESS') then ' latest major database version in use'
```
Comment (madhushreeray30) [inline query/sql.sp:925]:
```suggestion
       when (arguments ->> 'database_version') in ('POSTGRES_15', 'MYSQL_8_0', 'SQLSERVER_2019_STANDARD', 'SQLSERVER_2019_WEB', 'SQLSERVER_2019_ENTERPRISE', 'SQLSERVER_2019_EXPRESS') then ' latest major database version in use'
```

File: query/vertexai.sp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments -> 'encryption_spec' -> 'kms_key_name') is not null then 'ok'
+        else 'alarm'
+      end as status,
+      name || case
+        when (arguments -> 'encryption_spec' -> 'kms_key_name') is not null then ' encrypted with kms cmk'
+        else ' not encrypted with kms cmk'
```
Comment (madhushreeray30) [inline query/vertexai.sp:11]:
```suggestion
        when (arguments -> 'encryption_spec' -> 'kms_key_name') is not null then ' encrypted with KMS CMK'
        else ' not encrypted with KMS CMK'
```

**Review by madhushreeray30 — APPROVED** [review]

File: controls/redis.sp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  title       = "Redis instances should have auth enabled"
+  description = "Ensure Redis instance has AUTH enabled. This control is non-compliant if AUTH is disabled for Redis instance."
+  query       = query.redis_instance_auth_enabled
+
+  tags = local.redis_compliance_common_tags
+}
+
+control "redis_instance_encryption_in_transit_enabled" {
+  title       = "Redis instances encryption in transit should be enabled"
+  description = "Ensure Redis instance encryption in transit is enabled. This control is non-compliant if encryption in transit is disabled for Redis instance."
```
Comment (misraved) [inline controls/redis.sp:31]:
```suggestion
  description = "Ensure Redis instance encryption in transit is enabled. This control is non-compliant if encryption in transit is disabled for Redis instances."
```

File: query/redis.sp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments ->> 'auth_enabled') = 'true' then 'ok'
+        else 'alarm'
+      end as status,
+      name || case
+        when (arguments ->> 'auth_enabled') = 'true' then ' AUTH enabled'
+        else ' AUTH disabled'
```
Comment (misraved) [inline query/redis.sp:11]:
```suggestion
        when (arguments ->> 'auth_enabled') = 'true' then ' Auth enabled'
        else ' Auth disabled'
```

Comment (misraved) [reply query/redis.sp:11]: @Priyanka-Chatterjee-2000 @madhushreeray30 what kind of Auth is being checked over here?

**Review by misraved — COMMENTED** [review]
@madhushreeray30 @Priyanka-Chatterjee-2000 please take a look at the review comments. Thanks!!

**Review by misraved — APPROVED** [review]

**Pull request merged**